src/Features/Core/Portable/Snippets/AbstractSnippetProvider.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Snippets
+{
+    internal abstract class AbstractSnippetProvider : ISnippetProvider
+    {
+        protected SyntaxAnnotation _cursorAnnotation = new();
+        protected SyntaxAnnotation _reformatAnnotation = new();
+
+        // Enumerates all the cases in which a particular snippet should occur. 
+        protected abstract Task<bool> IsValidSnippetLocationAsync(Document document, int position, CancellationToken cancellationToken);
+        // Gets the localized string that is displayed in the Completion list
+        protected abstract string GetSnippetDisplayName();
+        // Generates the new snippet's TextChange that is being inserted into the document
```

Comment:
These comments should all be XML comments 😛

Reply:
(and the ones in the two interfaces that you committed after I wrote this comment!)